### PR TITLE
Add LstmEncoderV1 assembly

### PR DESCRIPTION
**Pull request by christophmluscher**
Adds LSTM encoder

depends on #66

### CODE REVIEW

File: i6_models/assemblies/lstm/lstm_v1.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@dataclass
+class LstmEncoderV1Config(ModelConfiguration):
+    input_dim: int
+    embed_dim: int
+    embed_dropout: float
+    lstm_layers_cfg: LstmBlockV1Config
+    lstm_dropout: float
+    init_args: Optional[Dict[str, Any]] = None
```
Comment (NeoLegends) [inline i6_models/assemblies/lstm/lstm_v1.py:22]: Can you add some docs on what can go in here?

File: i6_models/assemblies/lstm/lstm_v1.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    @classmethod
+    def from_dict(cls, model_cfg_dict: Dict):
```
Comment (NeoLegends) [inline i6_models/assemblies/lstm/lstm_v1.py:25]: I haven't seen these functions for other config classes. What is the reason they're needed here? Should we add generic "serialization" support to all other parameter dataclasses as well?

**Review by NeoLegends — COMMENTED** [review]
Just two minor Qs, rest LGTM!

File: i6_models/assemblies/lstm/lstm_v1.py (hunk 2 above)
Comment (christophmluscher) [reply i6_models/assemblies/lstm/lstm_v1.py:25]: I think this depends on how you import/call your model and how the params are then passed to the model.. 
I like to have a dataclass in order to set the params in my config. and I like to have a single attribute `cfg` instead of numerous attributes for the model..
But passing a dict instead of the dataclass allows me to write someting like this in the returnn config without the need to import the data classes:
```
get_model = __import__("functools").partial(
    __import__(
        "apptek_asr.users.cluescher.lib.returnn.pytorch_networks.lm.lstm.alex_variant_v6",
        fromlist=["Model"],
    ).Model,
    **{
        "model_config_dict": {
            "vocab_dim": 249352,
            "embed_dim": 128,
            "hidden_dim": 512,
            "n_lstm_layers": 2,
            "init_args": {
                "init_args_w": {"func": "normal", "arg": {"mean": 0.0, "std": 0.1}},
                "init_args_b": {"func": "normal", "arg": {"mean": 0.0, "std": 0.1}},
            },
            "num_noise": 25,
            "bias": True,
            "bottle_neck_dim": 512,
            "dropout": 0.2,
        }
    }
)
```
maybe needed is therefor a bit strong.. :D

File: i6_models/assemblies/lstm/lstm_v1.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        model_cfg_dict = model_cfg_dict.copy()
+        model_cfg_dict["lstm_layers_cfg"] = LstmBlockV1Config.from_dict(model_cfg_dict["lstm_layers_cfg"])
+        return cls(**model_cfg_dict)
+
+
+class LstmEncoderV1(nn.Module):
+    def __init__(self, model_cfg: Union[LstmEncoderV1Config, Dict], **kwargs):
+        super().__init__()
+
+        self.cfg = LstmEncoderV1Config.from_dict(model_cfg) if isinstance(model_cfg, Dict) else model_cfg
```
Comment (albertz) [inline i6_models/assemblies/lstm/lstm_v1.py:35]:
```suggestion
        self.cfg = LstmEncoderV1Config.from_dict(model_cfg) if isinstance(model_cfg, dict) else model_cfg
```
`Dict` is from `typing`, used to express K/V type of a `dict`.

File: i6_models/assemblies/lstm/lstm_v1.py (hunk 2 above)
Comment (albertz) [inline i6_models/assemblies/lstm/lstm_v1.py:25]:
```suggestion
    def from_dict(cls, model_cfg_dict: Dict[str, Any]):
```

File: i6_models/assemblies/lstm/lstm_v1.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+from i6_models.config import ModelConfiguration
+from i6_models.parts.lstm import LstmBlockV1Config, LstmBlockV1
+
+
+@dataclass
+class LstmEncoderV1Config(ModelConfiguration):
+    """
+    :param init_args: used to initialize parameters of modules, example:
+                      ```
```
Comment (albertz) [inline i6_models/assemblies/lstm/lstm_v1.py:19]: I think we usually indent by 4 on a continued doc line, or not?

File: i6_models/assemblies/lstm/lstm_v1.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+class LstmEncoderV1Config(ModelConfiguration):
+    """
+    :param init_args: used to initialize parameters of modules, example:
+        ```
+        {
+            "init_args_w": {"func": "normal", "arg": {"mean": 0.0, "std": 0.1}},
+            "init_args_b": {"func": "normal", "arg": {"mean": 0.0, "std": 0.1}},
+        }
+        ```
+    """
```
Comment (Atticus1806) [inline i6_models/assemblies/lstm/lstm_v1.py:25]: Again also please add the docs for the other parameters.

File: i6_models/assemblies/lstm/lstm_v1.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    @classmethod
+    def from_dict(cls, model_cfg_dict: Dict[str, Any]):
+        model_cfg_dict = model_cfg_dict.copy()
+        model_cfg_dict["lstm_layers_cfg"] = LstmBlockV1Config.from_dict(model_cfg_dict["lstm_layers_cfg"])
+        return cls(**model_cfg_dict)
+
+
+class LstmEncoderV1(nn.Module):
+    def __init__(self, model_cfg: Union[LstmEncoderV1Config, Dict[str, Any]], **kwargs):
```
Comment (Atticus1806) [inline i6_models/assemblies/lstm/lstm_v1.py:42]: Do we need kwargs here?

File: i6_models/assemblies/lstm/lstm_v1.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+        self.lstm_block = LstmBlockV1(self.cfg.lstm_layers_cfg)
+        self.lstm_dropout = nn.Dropout(self.cfg.lstm_dropout)
+
+        if self.cfg.init_args is not None:
+            self._param_init(**self.cfg.init_args)
+
+    def _param_init(self, init_args_w=None, init_args_b=None):
+        for m in self.modules():
+            for name, param in m.named_parameters():
```
Comment (Atticus1806) [inline i6_models/assemblies/lstm/lstm_v1.py:65]: I think I would be a bit more cautious here. This would fail with `KeyError` if given something that does not have `func` and `arg` in it. Maybe assert this beforehand? should we maybe even enforce it in the config above via post init?

**Review by Atticus1806 — CHANGES_REQUESTED** [review]

File: i6_models/assemblies/lstm/lstm_v1.py (hunk 7 above)
Comment (christophmluscher) [reply i6_models/assemblies/lstm/lstm_v1.py:71]: Added a post init to handle this.

**Review by Atticus1806 — APPROVED** [review]

File: i6_models/assemblies/lstm/lstm_v1.py (hunk 1 above)
Comment (JackTemaki) [reply i6_models/assemblies/lstm/lstm_v1.py:38]: I would not have a dict where flexibility is not needed. So maybe `init_args_w` and `init_args_b` directly separated? Or do you want this to be in this "loose" format to later extend it to separate LSTM and embedding parameters? Usually I would say more explicit is always better, because here you can easily add entries that do nothing and you will never detect the error.

Comment (JackTemaki) [reply i6_models/assemblies/lstm/lstm_v1.py:38]: Ah, I take it back, the dict is added as kwargs to _param_init, I think then it crashes if you have wrong entries. Well, not optimal but maybe fine then.

**Review by curufinwe — APPROVED** [review]
